src/typechecker/FStar.TypeChecker.TcTerm.fs

```diff
@@ -2635,7 +2744,23 @@ and tc_eqn scrutinee env branch
              Some e, g in
 
   (* 3. Check the branch *)
-  let branch_exp, c, g_branch = tc_term pat_env branch_exp in
+  let branch_exp, c, g_branch =
+    let branch_exp =  //ascribe with the return annotation, if it exists
+      match ret_opt with
+      | None -> branch_exp
+      | Some asc ->
+        asc
+        |> SS.subst_ascription [NT (scrutinee, norm_pat_exp)]
+        |> U.ascribe branch_exp in
```

Comment:
Rather than adding an ascription node and then removing it, couldn't we just set the Env.expected_typ? 

Some pros/cons:
+ That would mean we don't disturb the syntax of the term to typecheck it
+ We don't re-typecheck the annotated returns clause
-  But, currently, we only support Env.expected_typ rather than Env.expected_effect.  Can we add support for Env.expected_comp_type so that we can handle both cases of a returns annotation?

Reply:
Yes, agreed, would be great to have `Env.expected_comp`. As we discussed offline, we can take it up separately.